subsys/fs/ext2/ext2_impl.c

```diff
@@ -0,0 +1,1542 @@
+
+			/* insert zeros to the end of last block */
+			uint32_t old_block = old_size / block_size;
+			uint32_t start_off = old_size % block_size;
+			uint32_t to_write = MIN(new_size - old_size, block_size - start_off);
+
+			rc = ext2_fetch_inode_block(inode, old_block);
+			if (rc < 0) {
+				return rc;
+			}
+
+			memset(inode_current_block_mem(inode) + start_off, 0, to_write);
+			rc = ext2_commit_inode_block(inode);
+			if (rc < 0) {
+				return rc;
+			}
+		}
+
+		/* There is no need to zero rest of blocks because they will be automatically
+		 * treated as zero filled.
+		 */
+
+	} else {
+		/* First removed block is just the number of used blocks.
+		 * (We count blocks from zero hence its number is just number of used blocks.)
+		 */
+		uint32_t start_blk = used_blocks;
+		int64_t removed_blocks;
+
+		LOG_DBG("Inode trunc from blk: %d", start_blk);
+
+		/* Remove blocks starting with start_blk. */
+		removed_blocks = ext2_inode_remove_blocks(inode, start_blk);
+		if (removed_blocks < 0) {
+			return removed_blocks;
+		}
+
+		LOG_DBG("Removed blocks: %lld (%lld)",
+				removed_blocks, removed_blocks * (block_size / 512));
+		inode->i_blocks -= removed_blocks * (block_size / 512);
+	}
+
+	inode->i_size = new_size;
+
+	LOG_DBG("New inode size: %d (blocks: %d)", inode->i_size, inode->i_blocks);
+
+	rc = ext2_commit_inode(inode);
+	if (rc < 0) {
+		return rc;
+	}
+
+	return 0;
+}
+
+static int write_one_block(struct ext2_data *fs, struct ext2_block *b)
+{
+	int ret = 0;
+
+	if (!(b->flags & EXT2_BLOCK_ASSIGNED)) {
+		ret = ext2_assign_block_num(fs, b);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	ret = ext2_write_block(fs, b);
+	return ret;
+}
+
+int ext2_inode_sync(struct ext2_inode *inode)
+{
+	int ret;
+	struct ext2_data *fs = inode->i_fs;
+
+	for (int i = 0; i < 4; ++i) {
+		if (inode->blocks[i] == NULL) {
+			break;
+		}
+		ret = write_one_block(fs, inode->blocks[i]);
+		if (ret < 0) {
+			return ret;
+		}
+		ret = fs->sync(fs);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+	return 0;
+}
+
+int ext2_get_direntry(struct ext2_dir *dir, struct fs_dirent *ent)
+{
+	if (dir->d_off >= dir->d_inode->i_size) {
+		/* end of directory */
+		ent->name[0] = 0;
+		return 0;
+	}
+
+	struct ext2_data *fs = dir->d_inode->i_fs;
+
+	int rc, ret = 0;
+	uint32_t block = dir->d_off / fs->block_size;
+	uint32_t block_off = dir->d_off % fs->block_size;
+
+	LOG_DBG("Reading dir entry from block %d at offset %d", block, block_off);
+
+	rc = ext2_fetch_inode_block(dir->d_inode, block);
+	if (rc < 0) {
+		return rc;
+	}
+
+	struct ext2_disk_dentry *de =
+		(struct ext2_disk_dentry *)(inode_current_block_mem(dir->d_inode) + block_off);
+	struct ext2_inode *inode = NULL;
+
+	LOG_DBG("inode=%d name_len=%d rec_len=%d", de->de_inode, de->de_name_len, de->de_rec_len);
+
+	/* Get name of directory entry. */
+	uint32_t len = MIN(de->de_name_len, EXT2_MAX_FILE_NAME);
+
+	memcpy(ent->name, de->de_name, len);
+	ent->name[len] = '\0';
```

Comment:
If entry has name longer than allowed, shouldn't this be reported as error? If you allow this and, assuming you have allowed max len = 4, now "hell" and "hello" are the same entries.